### PR TITLE
Issue/hotfix/#63

### PR DESCRIPTION
**Pull request by mzxyz**
#63

### CODE REVIEW

File: core/peerset/src/lib.rs [hunk 1]
```diff
@@ -257,13 +257,18 @@ impl Peerset {
 	}
 
 	fn on_report_peer(&mut self, peer_id: PeerId, score_diff: i32) {
+		// Hotfix: for invulnerable peer connection, check whether the node peer is reserved
+		let reserved = self.data.get_priority_group(RESERVED_NODES).unwrap_or_default();
+		let isReservedPeer = reserved.contains(&peer_id);
```
Comment (cowboy-bebug) [inline core/peerset/src/lib.rs]: Could u change this to `is_reserved_peer`? (just to be consistent)

File: core/peerset/src/lib.rs [hunk 2]
```diff
@@ -257,13 +257,18 @@ impl Peerset {
...
+
 		// We want reputations to be up-to-date before adjusting them.
 		self.update_time();
 
 		match self.data.peer(&peer_id) {
 			peersstate::Peer::Connected(mut peer) => {
 				peer.add_reputation(score_diff);
-				if peer.reputation() < BANNED_THRESHOLD {
+				// Only check reputation of non-reserved nodes
+				if !isReservedPeer && peer.reputation() < BANNED_THRESHOLD {
```
Comment (cowboy-bebug) [inline core/peerset/src/lib.rs]: Could u change this to is_reserved_peer? (just to be consistent)

File: core/peerset/src/lib.rs [hunk 3]
```diff
@@ -646,5 +651,40 @@ mod tests {
...
+		let fut = futures::future::poll_fn(move |cx| {
+			// We need one polling for the message to be processed.
+			assert_eq!(Stream::poll_next(Pin::new(&mut peerset), cx), Poll::Pending);
+
+			// Reserved node should be accepted.
+			peerset.incoming(peer_id.clone(), IncomingIndex(1));
+			while let Poll::Ready(msg) = Stream::poll_next(Pin::new(&mut peerset), cx) {
+				assert_eq!(msg.unwrap(), Message::Accept(IncomingIndex(1)));
+			}
+
```
Comment (cowboy-bebug) [inline core/peerset/src/lib.rs:685]: I think we have to wait a bit to see if the reputation actually changes
https://github.com/cennznet/plug-blockchain/blob/00e67a4ac93b433fc765c9de9e9014add4cc4cfe/core/peerset/src/lib.rs#L637-L644

**Review by cowboy-bebug — CHANGES_REQUESTED** [review]
Nice! I have a couple of comments

File: core/peerset/src/lib.rs (hunk 3 above)
Comment (cowboy-bebug) [reply core/peerset/src/lib.rs:685]: ^ Actually ignore this
This is to wait for the reputation to come back up.

**Comment by jordy25519** [comment]
https://github.com/cennznet/plug-blockchain/blob/96c80bfa67f9b511b560d2468b355253e9f46a4c/core/peerset/src/lib.rs#L679
This line is a compile error. `peer_id` hasn't been declared change to`bootnode`

**Comment by cowboy-bebug** [comment]
I think it should be `reserved_peer`
```rust
 peerset.incoming(reserved_peer.clone(), IncomingIndex(1)); 
```

test fail with the above change 😆
```
---- tests::test_peerset_reserved_node_not_banned stdout ----
thread 'tests::test_peerset_reserved_node_not_banned' panicked at 'assertion failed: `(left == right)`
  left: `Ready(Some(Connect(PeerId("QmVEr8KUCkhAu26FQdBLempHqgkZheXHZsYPL9mBQT2Uq3"))))`,
 right: `Pending`', core/peerset/src/lib.rs:676:4
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

File: core/peerset/src/lib.rs [hunk 4]
```diff
@@ -646,5 +651,40 @@ mod tests {
...
+	#[test]
+	fn test_peerset_reserved_node_not_banned() {
+		let bootnode = PeerId::random();
+		let reserved_peer = PeerId::random();
+		let config = PeersetConfig {
+			in_peers: 0,
+			out_peers: 1,
+			bootnodes: vec![bootnode],
+			reserved_only: true,
+			reserved_nodes: Vec::new(),
```
Comment (cowboy-bebug) [inline core/peerset/src/lib.rs:667]: this might need to be
```rust
reserved_nodes: vec![reserved_peer],
```

File: core/peerset/src/lib.rs (hunk 1 above)
Comment (mzxyz) [reply core/peerset/src/lib.rs]: Nice catch, forgot to use as rust style

File: core/peerset/src/lib.rs (hunk 4 above)
Comment (mzxyz) [reply core/peerset/src/lib.rs:667]: Just refer to test case `test_peerset_add_reserved_peer`, not sure the difference between `handle.add_reserved_peer()` and `reserved_nodes: vec![reserved_peer]`

File: core/peerset/src/lib.rs [hunk 5]
```diff
@@ -257,13 +257,18 @@ impl Peerset {
 	}
 
 	fn on_report_peer(&mut self, peer_id: PeerId, score_diff: i32) {
+		// Hotfix: for invulnerable peer connection, check whether the node peer is reserved
+		let reserved = self.data.get_priority_group(RESERVED_NODES).unwrap_or_default();
```
Comment (jordy25519) [inline core/peerset/src/lib.rs]: More idiomatic way
```rust
		let is_reserved_peer = match self.data.get_priority_group(RESERVED_NODES) {
			Some(reserved) => reserved.contains(&peer_id),
			None => false,
		};
```

**Review by jordy25519 — APPROVED** [review]
LGTM! (assuming CI build passes)

File: core/peerset/src/lib.rs [hunk 6]
```diff
@@ -646,5 +653,40 @@ mod tests {
 
 		futures::executor::block_on(fut);
 	}
+
+
+	#[test]
+	fn test_peerset_reserved_node_not_banned() {
+		let bootnode = PeerId::random();
+		let reserved_peer = PeerId::random();
+		let config = PeersetConfig {
```
Comment (cowboy-bebug) [inline core/peerset/src/lib.rs:662]: Please add one more peer (we will test this later)
```rust
let banned_peer = PeerId::random();
```

File: core/peerset/src/lib.rs [hunk 7]
```diff
@@ -646,5 +653,40 @@ mod tests {
...
+			reserved_only: true,
+			reserved_nodes: Vec::new(),
+		};
+
+		let (mut peerset, handle) = Peerset::from_config(config);
+		handle.add_reserved_peer(reserved_peer.clone());
+
+		// Setting reputation of reserved node under the threshold.
+		handle.report_peer(reserved_peer.clone(), BANNED_THRESHOLD - 1);
+
```
Comment (cowboy-bebug) [inline core/peerset/src/lib.rs:675]: Add this to report the `peer`
```rust
handle.report_peer(banned_peer.clone(), BANNED_THRESHOLD - 1);
```

File: core/peerset/src/lib.rs [hunk 8]
```diff
@@ -646,5 +653,40 @@ mod tests {
...
+		let fut = futures::future::poll_fn(move |cx| {
+			// We need one polling for the message to be processed.
+			assert_eq!(Stream::poll_next(Pin::new(&mut peerset), cx), Poll::Pending);
```
Comment (cowboy-bebug) [inline core/peerset/src/lib.rs:678]: need to change to:
```rust
assert_eq!(
    Stream::poll_next(Pin::new(&mut peerset), cx),
    Poll::Ready(Some(Message::Connect(reserved_peer.clone())))
);
```

File: core/peerset/src/lib.rs [hunk 9]
```diff
@@ -646,5 +653,40 @@ mod tests {
...
+
+			// Reserved node should be accepted.
+			peerset.incoming(reserved_peer.clone(), IncomingIndex(1));
+			while let Poll::Ready(msg) = Stream::poll_next(Pin::new(&mut peerset), cx) {
+				assert_eq!(msg.unwrap(), Message::Accept(IncomingIndex(1)));
+			}
+
```
Comment (cowboy-bebug) [inline core/peerset/src/lib.rs:685]: Let's add this in to make sure that normal peers with bad reputation are actually rejected:
```rust
// Banned node should be rejected.
peerset.incoming(banned_peer.clone(), IncomingIndex(1));
while let Poll::Ready(msg) = Stream::poll_next(Pin::new(&mut peerset), cx) {
    assert_eq!(msg.unwrap(), Message::Reject(IncomingIndex(1)));
}
```

**Review by cowboy-bebug — CHANGES_REQUESTED** [review]
I have a few suggestions to (1) avoid the following test failure and (2) improve it:
```
running 1 test
test tests::test_peerset_reserved_node_not_banned ... FAILED

failures:

---- tests::test_peerset_reserved_node_not_banned stdout ----
thread 'tests::test_peerset_reserved_node_not_banned' panicked at 'assertion failed: `(left == right)`
  left: `Ready(Some(Connect(PeerId("QmXw7dHZEFo7vC48oqFpZgC7mcGbXfue13eqjh2ppag85q"))))`,
 right: `Pending`', core/peerset/src/lib.rs:678:4
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

**Pull request closed without merging**